File: MakeAlfaNumData.scala

// This function returns a "0" if the operand string can't be converted to a number; otherwise, the input is unaltered
import scala.util.control.Exception._
import org.apache.spark.SparkConf
import org.apache.spark.SparkContext
import org.apache.spark.mllib.linalg.DenseMatrix
import java.util.Random
import Array._
import java.io.PrintWriter
import java.io.IOException
import org.apache.spark.mllib.linalg.{Vector, Vectors}
import org.apache.spark.mllib.stat.{MultivariateStatisticalSummary, Statistics}

object replaceText  {
   def with0Str(s : String) : String = {
      if ((allCatch.opt(s.toDouble)).isDefined == true)
         return s
      else
         return ("0")
    }
}  // replaceText 

object makeNiceStr {

   def makeNiceStr(AoB : Array[Byte]) : String = 
   {
      val niceAoC = "ABCDEFGHJKLMNPQRSTUVWXYZ23456789"

      var AoC = new Array[Char](AoB.length)
      var myInt : Int = 0
      var c : Char = ' '
   
      for (i <- 0 to AoB.length - 1)
      {
         // AoB(i) = AoB(i) & 31
         // workaround for Scala bug
         myInt = AoB(i)
         myInt = myInt & 31
         // prevent first char numeric (to prevent all numeric labels!)	  
	 if ((i == 0) && (myInt > 23))
	 {
	    myInt = myInt - 24
         }
         c = niceAoC.charAt(myInt)
	 AoC(i) = c
      }
      return(AoC.mkString(""))
    }
}

object makeAlfaNumData
{
   val conf = new SparkConf().setAppName("makeAlfaNumData")
   val sc = new SparkContext(conf)  

   def fnMakeAlfaNumData(outputFileName : String, numRows : Int, numCols : Int, meanAndVarianceArrs : Double* )
   { 
   // Generate an array of normally distributed random numbers
   // Compute their statistics and write a .csv file with the generated data

      var meanAndVariancesArrLength = meanAndVarianceArrs.length
      var numColsTimes2 = 2*numCols

      if (meanAndVariancesArrLength != 2*numCols)
      {
         println("Input Error, number of mean and variances incorrect; should match numCols")
         println(s"numCols = $numCols, number of mean and variances = $meanAndVariancesArrLength (both put together); Should be $numColsTimes2" )
         println("Usage: makeData(outputFileName : String, numRows : Int, numCols : Int, meanAndVarianceArrs : Double* )")
      
         return
      }

      val rng = new Random

      println(); println("print out a few rows generated...."); println();

      var vID = new Array[Byte](13)

      for (i <- 0 to math.min(4, numRows - 1))
      {
         println()
         // print a 13 char visual ID
         scala.util.Random.nextBytes(vID)
         print(makeNiceStr.makeNiceStr(vID)); print("\t")  
         for (j <- 0 to math.max(0, numCols - 1))
         {
            print(scala.util.Random.nextGaussian())
	    print("\t")
	    scala.util.Random.nextBytes(vID)
            print(makeNiceStr.makeNiceStr(vID.dropRight(scala.util.Random.nextInt(10)))); print("\t")	
         }
      }	  
	  
      println(); println(".....")

      println();println(); println("using means and variances of:")

      var mean = new Array[Double](numCols)
      var variance = new Array[Double](numCols)

      for (j <- 0 to numCols - 1)
      {
         mean(j) = meanAndVarianceArrs(j);
         variance(j) = meanAndVarianceArrs(numCols+j)
      }

      println();println("means:");
      for (j <- 0 to numCols - 1)
      {
         print(mean(j)); print("\t")
      }
      println()

      println();println("variances:")
      for (j <- 0 to numCols - 1)
      {
         print(variance(j)); print("\t")
      }
      println();
      println()
   
      var stdDev = new Array[Double](numCols)

      for (j <- 0 to numCols - 1)
         stdDev(j) = Math.sqrt(variance(j))

      println("print out a few rows of transformed data" ); println();

      for (i <- 0 to math.min(4, numRows - 1))
      {
         println()
         scala.util.Random.nextBytes(vID)
         print(makeNiceStr.makeNiceStr(vID)); print("\t")  
         for (j <- 0 to math.max(0, numCols - 1))
         {
            print(scala.util.Random.nextGaussian()*stdDev(j) + mean(j))
            print("\t")
            // Add alfa fields
	    scala.util.Random.nextBytes(vID)
            print(makeNiceStr.makeNiceStr(vID.drop(scala.util.Random.nextInt(10)))); print("\t")	 
         }
      }
      println(); println(".....")


      val fileHandle = new PrintWriter(outputFileName)

      for (i <- 0 to numRows - 1)
      {

         scala.util.Random.nextBytes(vID)
         fileHandle.print(makeNiceStr.makeNiceStr(vID))
         fileHandle.print(",")
         if (numCols > 1)
         {
            for (j <- 0 to math.max(0 , numCols - 2))
            {
	       fileHandle.print(scala.util.Random.nextGaussian()*stdDev(j) + mean(j))
               fileHandle.print(",")
               scala.util.Random.nextBytes(vID)
               fileHandle.print(makeNiceStr.makeNiceStr(vID).dropRight(scala.util.Random.nextInt(10))); fileHandle.print(",")
            }
            fileHandle.println(scala.util.Random.nextGaussian()*stdDev(numCols-1) + mean(numCols-1))
         }
      }

      fileHandle.close()

      println();println(s"Wrote $outputFileName"); println()

      // read outputFile to a vector

      // outputFile will be RDD[String]

      val arrnFile = sc.textFile(outputFileName)

      println();println(s"Read $outputFileName");println();

      println("print out a few rows read from file");println()
      arrnFile.take(5).foreach(println); println()

      // parsedData will be org.apache.spark.rdd.RDD[org.apache.spark.mllib.linalg.Vector]

      val parsedData = arrnFile.map(s => Vectors.dense(s.split(',').map(replaceText.with0Str(_)).map(_.toDouble)))
      println();println("print out a few vectors after converting from strings");println()
      parsedData.take(5).foreach(println); println()

      val summary: MultivariateStatisticalSummary = Statistics.colStats(parsedData)

      println();println("print out summary statistics, mean and variance, for each column");println()
      println(summary.mean)
      println(summary.variance);println()

      println("summary.count"); println(summary.count); println()
      println("summary.max"); println(summary.max); println()
      println("summary.min"); println(summary.min); println()
      println("summary.normL1"); println(summary.normL1); println()
      println("summary.normL2"); println(summary.normL2); println()
      println("summary.numnonZeros"); println(summary.numNonzeros); println()

   } // fnMakeAlfaNumData()


      var rc = -1
  
      def main(args : Array[String]) = {

      try
      {     
         fnMakeAlfaNumData("/home/bsrsharma/work/scala/arran.csv", 10000,3, 1.2, 3.4, 5.6,  7.8, 9.10, 11.12)
      }
      catch
      {
         case ex: IOException  => println("makeAlfaNumData threw " + ex.toString)
      }
      finally
      {
         println("makeAlfaNumData finally ended")
      } 
   }  // main
}  // makeAlfaNumData

File: MakeAlfaNumData.java
import org.apache.spark.api.java.JavaSparkContext;
import org.apache.spark.api.java.JavaRDD;
import org.apache.spark.SparkConf;

import org.apache.spark.mllib.linalg.DenseMatrix;
import java.util.Random;

import org.apache.spark.mllib.linalg.Vector;
import org.apache.spark.mllib.linalg.Vectors;

import java.io.PrintWriter;

import org.apache.spark.mllib.stat.MultivariateStatisticalSummary;
import org.apache.spark.mllib.stat.Statistics;

import java.io.IOException;

import java.util.Arrays;
import org.apache.spark.mllib.linalg.Matrix;
import org.apache.spark.mllib.linalg.Matrices;
import org.apache.spark.sql.SparkSession;
import org.apache.spark.api.java.function.Function;
 

public final class MakeAlfaNumData
{

   public static String makeNiceStr (byte[] AoB) 
   {
      String niceAoC = "ABCDEFGHJKLMNPQRSTUVWXYZ23456789";

      char[] AoC = new char[AoB.length];
      int i;
      int myInt;
      char c;
      
   
      for (i = 0; i < AoB.length; i++)
      {
         // AoB[i] = AoB[i] & 31
         myInt = AoB[i];
         myInt = myInt & 31;
         // prevent first char numeric (to prevent all numeric labels!)	  
	 if ((i == 0) && (myInt > 23))
	 {
	    myInt = myInt - 24;
         }
         c = niceAoC.charAt(myInt);
	 AoC[i] = c;
      }
      return(String.valueOf(AoC));
    }
    
   public static void shiftAndScaleMat (int maxRow, int maxCol, double mean[], double stdDev[], DenseMatrix origData, double arrn[][] )
   {
   
      int i,j;

      for (j = 0; j < maxCol; j++)
         for (i = 0; i < maxRow; i++)
	 {
	    arrn[i][j] = mean[j] + stdDev[j]*origData.apply(i, j);
         }		  
   }
   
   static class ParseDouble implements Function<String[], double[]>
   {
      @Override
      public double[] call(String[] sA)
      {
         double dA[] = new double[sA.length];
         int j;
         
         for (j = 0; j < sA.length; j++)
         try
         {
            dA[j] = Double.parseDouble(sA[j]);
         }
         catch (NumberFormatException nFE)
         {
            dA[j] = 0.0;
         }                   
            
         return dA;
      }
   }   


   public static int fnMakeAlfaNumData(JavaSparkContext sc, String outputFileName, int numRows, int numCols, double meanAndVarianceArrs[])
   { 
   // Generate an array of normally distributed random numbers
   // Compute their statistics and write a .csv file with the generated data

      int rc = 0;
      int meanAndVariancesArrLength;
      int numColsTimes2;
      int i,j;
      
      Random rng = new Random();
      
      byte vID[] = new byte[16];
      int rblen;
      double arrn[][] = new double[numRows][numCols];
      
      meanAndVariancesArrLength = meanAndVarianceArrs.length;
      numColsTimes2 = 2*numCols;

      if (meanAndVariancesArrLength != 2*numCols)
      {
         System.out.println("Input Error, number of mean and variances incorrect; should match numCols");
         System.out.println("numCols = " + numCols + "number of mean and variances = " + meanAndVariancesArrLength + " (both put together); Should be numColsTimes2" );
         System.out.println("Usage: makeData(outputFileName : String, numRows : Int, numCols : Int, meanAndVarianceArrs : Double* )");
         return -1;
      }

      System.out.println("DenseMatrix randn( " + numRows + " , " + numCols + " )");

      double values[] = new double[numRows*numCols];
      
      DenseMatrix origData = new DenseMatrix(numRows, numCols, values);
      origData = origData.randn(numRows, numCols, rng);

      System.out.println("print out a few rows generated....");

      for (i = 0; i < Math.min(4, numRows); i++)
      {
         // print a 13 char visual ID
         rng.nextBytes(vID);
         System.out.print(makeNiceStr(vID) + "\t");
         for (j = 0; j < numCols; j++)
         {
            System.out.print(origData.apply(i,j) + "\t");
	    rng.nextBytes(vID);
	    rblen = rng.nextInt(10);
            System.out.print(makeNiceStr(vID).substring(0,6+rblen) + "\t");
         }
         System.out.println();
      }	  
	  
      System.out.println("using means and variances of:");

      double mean[] = new double[numCols];
      double variance[] = new double[numCols];

      for (j = 0; j < numCols; j++)
      {
         mean[j] = meanAndVarianceArrs[j];
         variance[j] = meanAndVarianceArrs[numCols+j];
      }

      System.out.println("means:");
      for (j = 0; j < numCols; j++)
      {
         System.out.print(mean[j] + " , ");
      }
      System.out.println();

      System.out.println("variances:");
      for (j = 0; j < numCols; j++)
      {
         System.out.print(variance[j] + " , ");
      }
      System.out.println();
   
      double stdDev[] = new double[numCols];

      for (j = 0; j < numCols; j++)
         stdDev[j] = Math.sqrt(variance[j]);

      shiftAndScaleMat(numRows, numCols, mean, stdDev, origData, arrn);

      System.out.println("print out a few rows of transformed data" );

      for (i = 0; i < Math.min(4, numRows); i++)
      {
         rng.nextBytes(vID);
         System.out.print(makeNiceStr(vID) + "\t");
         for (j = 0; j < numCols; j++)
         {
            System.out.print(arrn[i][j] + "\t");
	    rng.nextBytes(vID);
	    rblen = rng.nextInt(10);
            System.out.print(makeNiceStr(vID).substring(0,6+rblen) + "\t");
         }
         System.out.println();
      }

      try
      {
         PrintWriter fileHandle = new PrintWriter(outputFileName);


         for (i = 0; i < numRows; i++)
         {
            rng.nextBytes(vID);
            fileHandle.print(makeNiceStr(vID));
            fileHandle.print(",");
            if (numCols > 1)
            {
               for (j = 0; j < numCols - 1; j++)
               {
	          fileHandle.print(String.valueOf(arrn[i][j]));
                  fileHandle.print(",");
                  rng.nextBytes(vID);
	          rblen = rng.nextInt(10);
                  fileHandle.print(makeNiceStr(vID).substring(0,6+rblen)); fileHandle.print(",");
               }
               fileHandle.println(String.valueOf(arrn[i][numCols - 1]));
            }
         }
         fileHandle.close();
      }
      
      catch   (java.io.FileNotFoundException FNFE)
      {
         System.out.println("PrintWriter threw " + FNFE);
      }               

      System.out.println("Wrote " + outputFileName);

      JavaRDD<String> arrnFile = sc.textFile(outputFileName);

      System.out.println("Read " + outputFileName);

      System.out.println("print out a few rows read from file");
      
      System.out.println( arrnFile.take(5) );

      JavaRDD<String[]> parsedData = arrnFile.map(s -> s.split(","));
      
      System.out.println("print out a few vectors after converting from strings");
/*      
      for (String[] sA: parsedData.take(5))
      {
         for (String s: sA)
            System.out.print( s + "\t" );
         System.out.println();
      }             
*/      
      JavaRDD<double[]> parsedDataD = parsedData.map(new ParseDouble());
/*      
      for (double[] dA: parsedDataD.take(5))
      {
         for (double d: dA)
            System.out.print( d + "\t" );
         System.out.println();
      }
      
      i = 0;
      for (double[] dA: parsedDataD.take(numRows))
      {
         if ((dA[0] != 0.0) || (dA[2] != 0.0) || (dA[4] != 0.0))
           System.out.println(i+"\t"+dA[0]+"\t"+dA[1]+"\t"+dA[2]+"\t"+dA[3]+"\t"+dA[4]+"\t"+dA[5]);
         i++;           
      }
*/      
      JavaRDD<Vector> parsedDataV = parsedDataD.map(dA -> Vectors.dense(dA));
      
      System.out.println( parsedDataV.take(5) );
      
      MultivariateStatisticalSummary summary;
      
      summary = Statistics.colStats(parsedDataV.rdd());
      
      System.out.println("print out summary statistics, mean and variance, for each column");
      System.out.println(summary.mean());
      System.out.println(summary.variance());

      System.out.println("summary.count"); System.out.println(summary.count());
      System.out.println("summary.max"); System.out.println(summary.max());
      System.out.println("summary.min"); System.out.println(summary.min());
      System.out.println("summary.normL1"); System.out.println(summary.normL1());
      System.out.println("summary.normL2"); System.out.println(summary.normL2());
      System.out.println("summary.numnonZeros"); System.out.println(summary.numNonzeros());
      
      return rc;

   } // fnMakeAlfaNumData()


   int rc = -1;
  
   
   public static void main(String[] args)
   {
   
         SparkSession spark = SparkSession.builder().appName("makeAlfaNumData").getOrCreate();

         JavaSparkContext sc = new JavaSparkContext(spark.sparkContext());
         
         double meanAndVarianceArrs[] = { 1.2, 3.4, 5.6,  7.8, 9.10, 11.12 };

         fnMakeAlfaNumData(sc, "/home/bsrsharma/work/java/arran.csv", 10000,3, meanAndVarianceArrs);

         System.out.println("makeAlfaNumData finally ended");

         spark.stop();
         
   }  // main
}  // MakeAlfaNumData